File: ch5/5-12.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

const tabLen = 4
const maxLine = 1000

func main() {
	var ts rune = '\t'
	if len(os.Args) > 1 {
		ts = rune(os.Args[1][0])
	}

	for {
		l, n := getLine(maxLine)
		if n <= 0 {
			break
		}
		entab(l, n, ts)
		detab(l, n, ts)
		fmt.Printf("%s\n", l)
	}
}

func entab(s []byte, l int, ts rune) {
	var tabAdded bool
	for i := 0; i < l; i++ {
		if rune(s[i]) == ' ' && rune(s[i+1]) == ' ' && rune(s[i+2]) == ' ' && rune(s[i+3]) == ' ' {
			fmt.Printf("%c", ts)
			tabAdded = true
		} else if rune(s[i]) == ' ' {
			if !tabAdded {
				fmt.Printf("%c", rune(s[i]))
			}
		} else {
			tabAdded = false
			fmt.Printf("%c", rune(s[i]))
		}
	}
}

func detab(s []byte, l int, ts rune) {
	for i := 0; i < l; i++ {
		if rune(s[i]) == ts {
			s[i] = byte(' ')
			for j := l - 1; j >= i+1; j-- {
				s[j+(tabLen-1)] = s[j]
			}
			for k := 1; k < tabLen; k++ {
				s[i+k] = byte(' ')
			}
		}
	}
}

func getLine(lim int) ([]byte, int) {
	var c byte
	var i int
	s := make([]byte, lim)
	r := bufio.NewReader(os.Stdin)
	for i = 0; i < lim-1; i++ {
		c, err := r.ReadByte()
		if rune(c) == '\n' || err != nil {
			break
		}
		s[i] = c
	}
	if rune(c) == '\n' {
		s[i] = c
		i++
	}
	return s, i
}

File: kr/ch5/5-12.c
#include <stdio.h>

#define MAXLINE 1024
#define TABSIZE 4

int getline(char line[], int maxline);
void entab(char line[], char nline[], int len);
void detab(char line[], char nline[], int len);

int main(void)
{
    int len;
    char line[MAXLINE];
    char nline[MAXLINE];

    for (int i = 0; i < MAXLINE; i++) {
        line[i]  = 0;
        nline[i] = 0;
    }

    while ((len = getline(line, MAXLINE)) > 0) {
        if (len > 0) {
            entab(line, nline, len);
            detab(line, nline, len);
            printf("%s", line);
        }
    }

    return 0;
}

void entab(char line[], char nline[], int maxline)
{
    int i, j, ns;
    j  = 0;
    ns = 0;

    for (i = 0; i < maxline; ++i) {
        char c = line[i];

        if (c == ' ') {
            ns++;
        }

        if (c == ' ' && ns == TABSIZE) {
            ns         = 0;
            j          = j - (TABSIZE - 1);
            nline[j++] = '\t';
        } else {
            nline[j++] = c;
        }

        if (c != ' ') {
            ns = 0;
        }
    }

    nline[j] = '\0';
}

void detab(char line[], char nline[], int maxline)
{
    int i, j;

    j = 0;

    for (i = 0; i < maxline; ++i) {
        char c = line[i];

        if (c == '\t') {
            for (int k = 0; k < TABSIZE; ++k) {
                nline[j++] = ' ';
            }
        } else {
            nline[j++] = c;
        }
    }

    nline[j] = '\0';
}

int getline(char s[], int lim)
{
    int c, i;

    for (i = 0; i < lim - 1 && (c = getchar()) != EOF && c != '\n'; ++i) {
        s[i] = c;
    }

    if (c == '\n') {
        s[i] = c;
        ++i;
    }

    s[i] = '\0';

    return i;
}
